Clover/app/src/main/java/org/floens/chan/ui/controller/export/ImportExportSettingsController.java

```diff
@@ -0,0 +1,322 @@
+
+            group.add(new LinkSettingView(this,
+                    context.getString(R.string.export_settings),
+                    context.getString(R.string.export_settings_to_a_file),
+                    v -> onExportClicked()));
+
+            group.add(new LinkSettingView(this,
+                    context.getString(R.string.import_settings),
+                    context.getString(R.string.import_settings_from_a_file),
+                    v -> onImportClicked()));
+
+            groups.add(group);
+        }
+    }
+
+    private void onExportClicked() {
+        String state = Environment.getExternalStorageState();
+        if (!Environment.MEDIA_MOUNTED.equals(state)) {
+            showMessage(context.getString(R.string.error_external_storage_is_not_mounted));
+            return;
+        }
+
+        getPermissionHelper().requestPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, granted -> {
+            if (granted) {
+                if (presenter != null) {
+                    navigationController.presentController(loadingViewController);
+                    presenter.doExport(settingsFile);
+                }
+            } else {
+                getPermissionHelper().showPermissionRequiredDialog(context,
+                        context.getString(R.string.update_storage_permission_required_title),
+                        context.getString(R.string.storage_permission_required_to_export_settings),
+                        this::onExportClicked);
+            }
+        });
+    }
+
+    private void showCreateCloverDirectoryDialog() {
+        String state = Environment.getExternalStorageState();
+        if (!Environment.MEDIA_MOUNTED.equals(state)) {
+            showMessage(context.getString(R.string.error_external_storage_is_not_mounted));
+            return;
+        }
+
+        // if we already have the permission and the Clover directory already exists - do not show
+        // the dialog again
+        if (getPermissionHelper().hasPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+            if (settingsFile.getParentFile().exists()) {
+                return;
+            }
+        }
+
+        if (!AndroidUtils.isApi16()) {
+            // we can't request READ_EXTERNAL_STORAGE permission on devices with API level below 16
+            onPermissionGrantedForDirectoryCreation();
+            return;
+        }
+
+        // Ask the user's permission to check whether the Clover directory exists and create it if it doesn't
+        AlertDialog dialog = new AlertDialog.Builder(context)
+                .setTitle(context.getString(R.string.clover_directory_may_not_exist_title))
+                .setMessage(context.getString(R.string.clover_directory_may_not_exist_message))
+                .setPositiveButton(context.getString(R.string.create), (dialog1, which) -> {
+                    getPermissionHelper().requestPermission(Manifest.permission.READ_EXTERNAL_STORAGE, granted -> {
```

Comment:
This should be WRITE_EXTERNAL_STORAGE, because we're making the directories here which is writing, not reading.

Reply:
Huh, that's funny, because couple of lines above I'm checking whether we already have WRITE_EXTERNAL_STORAGE and if not then for some reason requesting the READ_EXTERNAL_STORAGE.